sw/device/lib/crypto/api.h

```diff
@@ -0,0 +1,1378 @@
+  kEddsaSignModeEdDSA = 0,
+  // Signature mode Hashed EdDSA.
+  kEddsaSignModeHashEdDSA = 1
+} eddsa_sign_mode_t;
+
+/**
+ * Struct to handle ECC public key type.
+ *
+ * Length of public key coordinates (x,y) follows len(prime_P).
+ */
+typedef struct ecc_public_key {
+  // ECC public key x coordinate.
+  uint32_t *pubKey_x;
+  // ECC public key y coordinate.
+  uint32_t *pubKey_y;
+} ecc_public_key_t;
+
+/**
+ * Struct to handle EdDSA public key type.
+ *
+ * Length of encoded public key for ed25519 is 256 bits (32 bytes).
+ */
+typedef struct eddsa_public_key {
+  // Public key for ed25519
+  uint32_t *pubKey;
+} eddsa_public_key_t;
+
+/**
+ * Struct for domain parameters for a generic ECC Weierstrass curve.
+ */
+typedef struct ecc_domain {
+  // Length of prime P.
+  size_t len_prime_P;
+  // Value of prime P.
+  uint32_t *prime_P;
+  // Length of coefficient A.
+  size_t len_A;
+  // Value of coefficient A.
+  uint32_t *A;
+  // Length of coefficient B.
+  size_t len_B;
+  // Value of coefficient B.
+  uint32_t *B;
+  // Value of basepoint G’s x coordinate.
+  uint32_t *base_point_G_x;
+  // Value of basepoint G’s y coordinate.
+  uint32_t *base_point_G_y;
+  // Length of prime order of group.
+  size_t len_prime_order_Q;
+  // Value of prime order of group.
+  uint32_t *prime_order_Q;
+  // Cofactor of the curve.
+  uint32_t cofactor;
+  // Checksum value of the parameters.
+  uint32_t checksum;
+} ecc_domain_t;
+
+/**
+ * Struct to handle crypto data buffer with pointer and length.
+ */
+typedef struct crypto_uint8_buf {
+  // Pointer to the data.
+  uint8_t *data;
+  // Length of the data in bytes.
+  size_t len;
+} crypto_uint8_buf_t;
+
+/**
+ * Performs the AES initialization operation.
+ *
+ * The `aes_mode`and `aes_operation` are selected and the `key`, `iv`
+ * are loaded into the register.
+ *
+ * @param key Pointer to the blinded key struct with key shares
+ * @param iv Initialization vector, used for CBC, CFB, OFB, CTR modes
+ * @param aes_mode Required aes mode of operation
+ * @param aes_operation Required aes operation
+ * @return crypto_status_t The result of the init operation
+ */
+crypto_status_t aes_init(const crypto_blinded_key_t *key, crypto_uint8_buf_t iv,
+                         block_cipher_mode_t aes_mode,
+                         aes_operation_t aes_operation);
+
+/**
+ * Performs the AES cipher operation.
+ *
+ * The #aes_init function should be called before this, to initialize
+ * the key, aes mode and aes cipher operation to be performed.
+ *
+ * The input data in the `cipher_input` is first padded using the
+ * aes_padding scheme, the output is placed in the `cipher_output`.
+ *
+ * @param cipher_input Input data to be ciphered
+ * @param aes_padding Padding scheme to be used for the data
+ * @param cipher_output Output data after cipher operation
+ * @return crypto_status_t The result of the cipher operation
+ */
+crypto_status_t aes_cipher(const crypto_uint8_buf_t cipher_input,
```

Comment:
ISTR that we proposed putting a prefix like `otcrypto_` in front of every name. Did that ever materialize?

Reply:
Not as far as I know; I've not heard this suggestion before.

Reply:
I'm going to comment on this in the doc. I might have said it in a meeting and just never wrote it down. A lot of the names in this library are extremely short and those are a bad idea from a linking perspective.